src/Contracts/OpdexVault.cs

```diff
@@ -81,129 +136,348 @@ public void NotifyDistribution(UInt256 amount)
-        var vestingDuration = VestingDuration;
+        ValidateProposal(amount, description);
 
-        Assert(Message.Sender == owner, "OPDEX: UNAUTHORIZED");
-        Assert(to != owner, "OPDEX: INVALID_CERTIFICATE_HOLDER");
-        Assert(amount > 0 && amount <= TotalSupply, "OPDEX: INVALID_AMOUNT");
-        Assert(Block.Number < Genesis + vestingDuration, "OPDEX: TOKENS_BURNED");
+        var certificate = GetCertificate(recipient);
 
-        var certificates = GetCertificates(to);
+        Assert(certificate.VestedBlock == 0ul, "OPDEX: CERTIFICATE_EXISTS");
 
-        Assert(certificates.Length < MaximumCertificates, "OPDEX: CERTIFICATE_LIMIT_REACHED");
+        var totalProposedAmount = TotalProposedAmount;
 
-        var vestedBlock = Block.Number + vestingDuration;
+        Assert(TotalSupply - totalProposedAmount >= amount, "OPDEX: INSUFFICIENT_VAULT_SUPPLY");
 
-        certificates = InsertCertificate(certificates, amount, vestedBlock, false);
+        TotalProposedAmount = totalProposedAmount + amount;
 
-        SetCertificates(to, certificates);
+        return CreateCertificateTypeProposalExecute(recipient, amount, description, (byte)ProposalType.Create);
+    }
 
-        TotalSupply -= amount;
+    /// <inheritdoc />
+    public ulong CreateRevokeCertificateProposal(Address recipient, string description)
+    {
+        var certificate = GetCertificate(recipient);
+        var maxExpiration = Block.Number + VoteDuration + PledgeDuration;
 
-        Log(new CreateVaultCertificateLog{ Owner = to, Amount = amount, VestedBlock = vestedBlock });
+        Assert(!certificate.Revoked && certificate.VestedBlock > maxExpiration, "OPDEX: INVALID_CERTIFICATE");
+        ValidateProposal(certificate.Amount, description);
+        return CreateCertificateTypeProposalExecute(recipient, certificate.Amount, description, (byte)ProposalType.Revoke);
     }
 
     /// <inheritdoc />
-    public void RedeemCertificates()
+    public ulong CreatePledgeMinimumProposal(UInt256 amount, string description)
     {
-        var certificates = GetCertificates(Message.Sender);
-        var lockedCertificates = new VaultCertificate[0];
-        var amountToTransfer = UInt256.Zero;
+        ValidateProposal(amount, description);
+        return CreateMinimumAmountChangeProposalExecute(amount, description, (byte)ProposalType.PledgeMinimum);
+    }
 
-        foreach (var certificate in certificates)
-        {
-            if (certificate.VestedBlock > Block.Number)
-            {
-                lockedCertificates = InsertCertificate(lockedCertificates, certificate.Amount, certificate.VestedBlock, certificate.Revoked);
-                continue;
-            }
+    /// <inheritdoc />
+    public ulong CreateProposalMinimumProposal(UInt256 amount, string description)
+    {
+        ValidateProposal(amount, description);
+        return CreateMinimumAmountChangeProposalExecute(amount, description, (byte)ProposalType.ProposalMinimum);
+    }
```

Comment:
In a scenario where the pledge/vote minimum is believed to be too low and a proposal to made to increase this threshold, it may be worth requiring the proposal to reach the pledge/vote minimum that is being proposed.

Reply:
They do, minimums are checked to validate a vote at the very end when deciding if the vote was approved or not. It is not evaluated at all during creation here.